### ISSUE TITLE
Time of Day and Holiday Routing - "mweek" not matching valid dates

### ISSUE THREAD

**Issue by IainKay** [signalwire/freeswitch#313]
An example in my dial plan:

        <extension name="TEST-RULE" continue="true">
                <condition wday="4" mweek="1" time-of-day="22:30-23:30" break="never">
                        <action application="set" data="status=closed" inline="true"/>
                </condition>
        </extension>

Today **05/02/2020** at the time of **23:08** I placed a test call and this is what is logged:

        a4d9483c-87ec-4d8f-adad-af3d7c6bed0a Dialplan: sofia/internal/1000@sip.kay.family Date/TimeMatch (FAIL) [TEST-RULE] break=never

Looking at the calendar I can see that **05/02/2020** is the **first** (**mweek=1**) **Wednesday** (**wday=4**) and the **time-of-day** is between **22:30-23:30**.

If, however, I change the dialplan entry to the below:

        <extension name="TEST-RULE" continue="true">
                <condition wday="4" mweek="2" time-of-day="22:30-23:30" break="never">
                        <action application="set" data="status=closed" inline="true"/>
                </condition>
        </extension>

This then passes as logged below:

        06172734-7ba9-4dc6-99d9-772a06e07c8a Dialplan: sofia/internal/1000@sip.kay.family Date/Time Match (PASS) [TEST-RULE] break=never

It appears that FreeSWITCH is considering 05/02/2020 to be the _second_ Wednesday of this month.

Freeswitch version:

    1.10.2~release~13~f7bdd3845a~buster-1~buster+1

**Comment by IainKay** [comment]
Increasing core debug-level to 10 yields slightly more information:

    Dialplan: sofia/internal/1000@81.161.226.252 parsing [default->TEST-RULE] continue=true
    2020-02-06 09:12:05.470008 [ALERT] switch_xml.c:3205 XML DateTime: week of month[2] =~ 1 (FAIL)
    Dialplan: sofia/internal/1000@81.161.226.252 Date/TimeMatch (FAIL) [TEST-RULE] break=never

Interestingly other time based rules don't seem to be applying correctly on my box as I see that the day is identified as Friday today:

    Dialplan: sofia/internal/1000@81.161.226.252 parsing [default->7000 - Time based rule - Mon-Fri 9am-5pm] continue=true
    2020-02-06 09:12:05.470008 [ALERT] switch_xml.c:3212 XML DateTime Check: day of week[fri] =~ 2-6 (PASS)
    2020-02-06 09:12:05.470008 [ALERT] switch_xml.c:3241 XML DateTime Check: time of day[9:12:5] =~ 09:00-17:00 (PASS)
    Dialplan: sofia/internal/1000@81.161.226.252 Date/Time Match (PASS) [7000 - Time based rule - Mon-Fri 9am-5pm] break=never

But it's Thursday:

    root@freeswitch-01:/home/iainkay# date
    Thu  6 Feb 09:21:24 GMT 2020

System timezone is Europe/London.

FreeSWITCH timezone is Europe/London as set in /etc/freeswitch/vars.xml:

    <X-PRE-PROCESS cmd="set" data="timezone=Europe/London"/>

**Comment by StefanYohansson** [comment]
Inspecting `switch_xml.c` seems that we have a bug probably related to when leap year.

Basically, from line 3201 we have

```
if (time_match && xmweek) {
               /* calculate the day of the week of the first of the month (0-6) */
               int firstdow = (int) (7 - (tm.tm_mday - (tm.tm_wday + 1)) % 7) % 7;
               /* calculate the week of the month (1-6)*/
               int test = (int) ceil((tm.tm_mday + firstdow) / 7.0);
               time_match = switch_number_cmp(xmweek, test);
               switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_DEBUG9,
                               "XML DateTime: week of month[%d] =~ %s (%s)\n", test, xmweek, time_match ? "PASS" : "FAIL");
       }
```

while executing FS with server set for your issue case date, we have that firstdow (first day of week) is 6 (saturday) but I think first day of week is Feb 2 on Sunday, no?

When code continues execution, I see that `ceil((tm.tm_mday + firstdow) / 7.0);` results in 2 because (5 + 6) = 11 / 7 =~ 1.5 than on ceil it turns 2.

It's tricky because Feb indeed has a "first week" only with day 1 that is a Saturday (why the firstdow is 6, i guess). But I think it doesnt match the logic that needs to be something like `the first (week 1) wednesday (4) of the month`

I am digging on it in order to see if I can fix it.

**Comment by IainKay** [comment]
> Inspecting switch_xml.c seems that we have a bug probably related to when leap year.

Interesting. I did not consider the possibility that the leap year may be throwing things off.

Since posting my bug I had a look at the code myself, adding some extra debug output to get more info, but have not made any headway there.

> while executing FS with server set for your issue case date, we have that firstdow (first day of week) is 6 (saturday) but I think first day of given week is Feb 2 on Sunday, no?

Afaik officially the UK (Europe/London timezone) adopted ISO 8601 making Monday the first day of the week. On that basis I would expect a Europe/London timezone to find the first dow as Monday 03/02/2020.

Something like American/New_York I would expect Sunday 02/02/2020.

**Comment by ydinari** [comment]
Hi, isn't that expected, since mweek by definition is NOT equal to the N-th weekday in the month, just like in the example that was given. In that example, of February 2020, the first week doesn't contain a Wednesday, so the first Wednesday happens on the second week (mweek=2).
Shouldn't there be another parameter, that specifies the N-th occurrence of some weekday in the month?
That parameter range will be by the way 1-5 and not 1-6, as no day can happen 6 times in a month.

**Comment by StefanYohansson** [comment]
Yeah, I think a new parameter could solve the problem but IMHO if freeswitch goes the way to create a new thing for that, they should consider implementing [Recurrence Rule RFC](https://tools.ietf.org/html/rfc5545) that should solve any kind of recurrence problem.

a simple natural language parameter with value "every month on the 1nd Wednesday" or "RRULE:FREQ=MONTHLY;BYDAY=+1WE" for whoever likes the hard way.

I have another project using RRULE right now, maybe when I have some spare time I can try to insert RRULE support on freeswitch.

**Comment by IainKay** [comment]
> Hi, isn't that expected, since mweek by definition is NOT equal to the N-th weekday in the month, just like in the example that was given. In that example, of February 2020, the first week doesn't contain a Wednesday, so the first Wednesday happens on the second week (mweek=2).

This isn't the impression I got when reading the documentation.

_The documentation provides the following example:_

`<!-- president's day is the 3rd Monday in February -->`
`<condition wday="2" mweek="3" mon="2">`

That to me would suggest that the **mweek** variable _is_ the N-th weekday in the month.

Reference: https://freeswitch.org/confluence/display/FREESWITCH/Time+of+Day+and+Holiday+Routing#TimeofDayandHolidayRouting-HolidayRouting


So either the documentation is wrong or there's a bug with the **mweek** variable.

Based on what has been said, it sounds like the documentation should be updated to remove this example.

Perhaps it could be re-written as:

> `<!-- president's day is the 3rd Monday in February -->`
> `<condition wday="2" mon="2" mday="15-21">`

**Comment by ydinari** [comment]
I read that example too (https://freeswitch.org/confluence/display/FREESWITCH/Time+of+Day+and+Holiday+Routing#TimeofDayandHolidayRouting-HolidayRouting), but I also looked at the code - https://github.com/signalwire/freeswitch/blob/53cd06900c5ce827b2c0ac74485607d9a07d4a7d/src/switch_xml.c#L3198,
and I think the description of mweek is right, and the example in the documentation is wrong.
